examples/ode.py

```diff
@@ -0,0 +1,112 @@
+(prey) populations. We will use the dataset obtained from [3] and run MCMC to get
+inferences about parameters of the differential equation governing the dynamics.
+
+**References:**
+
+    1. Bob Carpenter (2018), `"Predator-Prey Population Dynamics: the Lotka-Volterra model in Stan"
+       <https://mc-stan.org/users/documentation/case-studies/lotka-volterra-predator-prey.html/>`_.
+    2. https://en.wikipedia.org/wiki/Lotka–Volterra_equations
+    3. http://people.whitman.edu/~hundledr/courses/M250F03/M250.html
+"""
+
+import argparse
+import os
+
+import matplotlib
+import matplotlib.pyplot as plt
+
+from jax.experimental.ode import build_odeint
+import jax.numpy as np
+from jax.random import PRNGKey
+
+import numpyro
+import numpyro.distributions as dist
+from numpyro.examples.datasets import LYNXHARE, load_dataset
+from numpyro.infer import MCMC, NUTS, Predictive
+
+matplotlib.use('Agg')  # noqa: E402
+
+
+def dz_dt(z, t, alpha, beta, gamma, delta):
+    """
+    Lotka–Volterra equations. Real positive parameters `alpha`, `beta`, `gamma`, `delta`
+    describes the interaction of two species.
+    """
+    u = z[0]
+    v = z[1]
+    du_dt = (alpha - beta * v) * u
+    dv_dt = (-gamma + delta * u) * v
+    return np.stack([du_dt, dv_dt])
+
+
+predator_prey_int = build_odeint(dz_dt, rtol=1e-5, atol=1e-3, mxstep=500)
+
+
+def model(N, y=None):
+    """
+    :param int N: number of measurement times
+    :param numpy.ndarray y: measured populations with shape (N, 2)
+    """
+    # initial population
+    z_init = numpyro.sample("z_init", dist.LogNormal(np.log(10), 1), sample_shape=(2,))
+    # measurement times
+    ts = np.arange(float(N))
+    # parameters alpha, beta, gamma, delta of dz_dt
+    theta = numpyro.sample(
+        "theta",
+        dist.TruncatedNormal(low=0., loc=np.array([0.5, 0.05, 1.5, 0.05]),
+                             scale=np.array([0.5, 0.05, 0.5, 0.05])))
+    # integrate dz/dt, the result will have shape N x 2
+    z = predator_prey_int(z_init, ts, *theta)
+    # measurement errors, we expect that measured hare has larger error than measured lynx
+    sigma = numpyro.sample("sigma", dist.Exponential(np.array([1, 2])))
+    # measured populations (in log scale)
+    numpyro.sample("y", dist.Normal(np.log(z), sigma), obs=y)
+
+
+def main(args):
+    _, fetch = load_dataset(LYNXHARE, shuffle=False)
+    year, data = fetch()  # data is in hare -> lynx order
+
+    # use dense_mass for better mixing rate
+    mcmc = MCMC(NUTS(model, dense_mass=True),
+                args.num_warmup, args.num_samples, num_chains=args.num_chains,
+                progress_bar=False if "NUMPYRO_SPHINXBUILD" in os.environ else True)
+    mcmc.run(PRNGKey(1), N=data.shape[0], y=np.log(data))
+    mcmc.print_summary()
+
+    # predict populations
+    y_pred = Predictive(model, mcmc.get_samples())(PRNGKey(2), data.shape[0])["y"]
+    pop_pred = np.exp(y_pred)
+    plt.fill_between(year, data[:, 0], color="y", alpha=0.2, label="true hare")
```

Comment:
Could you explain why are we filling in here (I thought these values are fixed measurements)? Also, could you add the predictive CI to the plots?

Reply:
I just make sure that the plot is easy to observe: https://github.com/pyro-ppl/numpyro/files/4006718/ode_plot1.pdf. Let me see if adding CI is helpful for the plot.

Reply:
I would suggest using a different line style. Fill makes it seem like it could be anywhere between 0 to actual value. We should use a fill for the CI though (I think that will be nice to have as a posterior predictive check).

Reply:
Actually, I mimic this figure :D https://en.wikipedia.org/wiki/Lotka%E2%80%93Volterra_equations#/media/File:Milliers_fourrures_vendues_en_environ_90_ans_odum_1953_en.jpg

Reply:
![Screenshot from 2020-01-14 17-15-12](https://user-images.githubusercontent.com/4736342/72387403-77042500-36f1-11ea-9a74-ae36411de9ca.png)
This is the plot with 60% CI. It is hard to appreciate hare's PI. Should I remove it? I want to have hare and lynx in the same plot to illustrate predator-prey pattern.

Reply:
I think it might look nicer if we replace the true hare population with points or a line and use a light shade for the CI (that's kind of the point of using numpyro instead of just getting point estimates, so will be good to have in the figure). Feel free to commit your changes, and I can try to play around with the plot settings too.

Reply:
yeah, let me see if I can come up with something cleaner with CI. :)

Reply:
![Screenshot from 2020-01-15 11-16-15](https://user-images.githubusercontent.com/4736342/72450519-84b8b980-3788-11ea-9f1a-5cc04546712a.png)

It looks much better now. Thanks for your suggestions!

Reply:
Looks great, thanks!